test/test_spectral_ops.py

```diff
@@ -43,57 +47,162 @@ def method_fn(t):
+    @precisionOverride({torch.complex64: 1e-4, torch.float: 1e-4})
+    @dtypes(torch.float, torch.double, torch.complex64, torch.complex128)
+    def test_fft_numpy(self, device, dtype):
+        test_args = product(
+            # input
+            (torch.randn(67, device=device, dtype=dtype),
+             torch.randn(80, device=device, dtype=dtype),
+             torch.randn(12, 14, device=device, dtype=dtype),
+             torch.randn(9, 6, 3, device=device, dtype=dtype)),
+            # n
+            (None, 50, 6),
+            # dim
+            (-1, 0),
+            # norm
+            ((None, "forward", "backward", "ortho")
+             if LooseVersion(np.__version__) >= '1.20.0'
+             else (None, "ortho"))
+        )
+
+        fft_functions = ['fft', 'ifft', 'hfft', 'irfft']
+        # Real-only functions
+        if not dtype.is_complex:
+            fft_functions += ['rfft', 'ihfft']
+
+        for fname in fft_functions:
+            torch_fn = getattr(torch.fft, fname)
+            numpy_fn = getattr(np.fft, fname)
+
+            def fn(t: torch.Tensor, n: Optional[int], dim: int, norm: Optional[str]):
+                return torch_fn(t, n, dim, norm)
+            scripted_fn = torch.jit.script(fn)
+
+            # TODO: revisit the following function if t.fft() becomes torch.fft.fft
+            # def method_fn(t, n, dim, norm):
+            #     return getattr(t, fname)(n, dim, norm)
+            # scripted_method_fn = torch.jit.script(method_fn)
+
+            # TODO: revisit the following function if t.fft() becomes torch.fft.fft
+            # torch_fns = (torch.fft.fft, torch.Tensor.fft, scripted_fn, scripted_method_fn)
+            torch_fns = (torch_fn, scripted_fn)
+
+            for iargs in test_args:
+                args = list(iargs)
+                input = args[0]
+                args = args[1:]
+
+                expected = numpy_fn(input.cpu().numpy(), *args)
+                exact_dtype = dtype in (torch.double, torch.complex128)
+                for fn in torch_fns:
+                    actual = fn(input, *args)
+                    self.assertEqual(actual, expected, exact_dtype=exact_dtype)
 
     # Note: NumPy will throw a ValueError for an empty input
     @skipCUDAIfRocm
     @skipCPUIfNoMkl
     @onlyOnCPUAndCUDA
-    @dtypes(torch.complex64, torch.complex128)
+    @dtypes(torch.float, torch.double, torch.complex64, torch.complex128)
     def test_empty_fft(self, device, dtype):
         t = torch.empty(0, device=device, dtype=dtype)
+        match = r"Invalid number of data points \([-\d]*\) specified"
 
-        if self.device_type == 'cuda':
-            with self.assertRaisesRegex(RuntimeError, "cuFFT error"):
-                torch.fft.fft(t)
-            return
-
-        # CPU (MKL)
-        with self.assertRaisesRegex(RuntimeError, "MKL FFT error"):
+        with self.assertRaisesRegex(RuntimeError, match):
```

Comment:
Nit: you can code golf this by iterating over the ops:

```
for op in (torch.fft.fft, torch.fft.ifft, ...):
```

You'd have to expand the op list based on the dtype like you do in the previous test to handle rfft and ihfft.

Reply:
Done.